vcd/resource_vcd_network_routed.go

```diff
@@ -227,41 +238,156 @@ func resourceVcdNetworkRoutedCreate(d *schema.ResourceData, meta interface{}) er
 
 	}
 
-	d.SetId(d.Get("name").(string))
+	d.SetId(network.OrgVDCNetwork.ID)
 
-	return resourceVcdNetworkRead(d, meta)
+	return resourceVcdNetworkRoutedRead(d, meta)
 }
 
-func resourceVcdNetworkRead(d *schema.ResourceData, meta interface{}) error {
+func resourceVcdNetworkRoutedRead(d *schema.ResourceData, meta interface{}) error {
 	vcdClient := meta.(*VCDClient)
 
 	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
 	if err != nil {
 		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
 	}
+	identifier := d.Id()
 
-	network, err := vdc.FindVDCNetwork(d.Id())
+	if identifier == "" {
+		identifier = d.Get("name").(string)
+	}
+	network, err := vdc.GetOrgVdcNetworkByNameOrId(identifier, false)
 	if err != nil {
-		log.Printf("[DEBUG] Network no longer exists. Removing from tfstate")
-		d.SetId("")
-		return nil
+		log.Printf("[DEBUG] Network %s no longer exists. Removing from tfstate", identifier)
+		return fmt.Errorf("[network routed read] error retrieving Org VDC network %s: %s", identifier, err)
 	}
+	edgeGatewayName := d.Get("edge_gateway").(string)
 
-	d.Set("name", network.OrgVDCNetwork.Name)
-	d.Set("href", network.OrgVDCNetwork.HREF)
+	// When this function is called from the data source
+	if edgeGatewayName == "" {
+		edgeGatewayName, err = findEdgeGatewayConnection(vcdClient, vdc, network)
+		if err != nil {
+			return fmt.Errorf("[network_routed read] no edge gateway connection found for network %s: %s", network.OrgVDCNetwork.Name, err)
+		}
+		_ = d.Set("edge_gateway", edgeGatewayName)
+	}
+	edgeGateway, err := vdc.GetEdgeGatewayByName(edgeGatewayName, false)
+	if err != nil {
+		log.Printf("[DEBUG] error retrieving edge gateway")
+		return fmt.Errorf("[network routed read] error retrieving edge gateway %s: %s", edgeGatewayName, err)
+	}
+
+	_ = d.Set("name", network.OrgVDCNetwork.Name)
+	_ = d.Set("href", network.OrgVDCNetwork.HREF)
+	_ = d.Set("shared", network.OrgVDCNetwork.IsShared)
 	if c := network.OrgVDCNetwork.Configuration; c != nil {
-		d.Set("fence_mode", c.FenceMode)
+		_ = d.Set("fence_mode", c.FenceMode)
 		if c.IPScopes != nil {
-			d.Set("gateway", c.IPScopes.IPScope[0].Gateway)
-			d.Set("netmask", c.IPScopes.IPScope[0].Netmask)
-			d.Set("dns1", c.IPScopes.IPScope[0].DNS1)
-			d.Set("dns2", c.IPScopes.IPScope[0].DNS2)
+			_ = d.Set("gateway", c.IPScopes.IPScope[0].Gateway)
+			_ = d.Set("netmask", c.IPScopes.IPScope[0].Netmask)
+			_ = d.Set("dns1", c.IPScopes.IPScope[0].DNS1)
+			_ = d.Set("dns2", c.IPScopes.IPScope[0].DNS2)
+			_ = d.Set("dns_suffix", c.IPScopes.IPScope[0].DNSSuffix)
 		}
 	}
 
+	dhcp := getDhcpFromEdgeGateway(network.OrgVDCNetwork.HREF, edgeGateway)
```

Comment:
These two `pool` set code looks like duplication - same is in previous resource.

Reply:
If you refer to network_isolated, they are two different things. One is internal DHCP (network_isolated) and the other gets DHCP from the edge gateway (network_routed)

Reply:
```

Reply:
I can find a way to reuse them, but then it would be harder to read.

Reply:
that's fine then